pkg/operator/ceph/object/admin.go

```diff
@@ -292,6 +295,89 @@ func isInvalidFlagError(err error) bool {
 	return exitCode == 22
 }
 
+// CommitConfigChanges commits changes to RGW configs for realm/zonegroup/zone changes idempotently.
+// Under the hood, this updates the RGW config period and commits the change if changes are detected.
+func CommitConfigChanges(c *Context) error {
+	currentPeriod, err := runAdminCommand(c, true, "period", "get")
+	if err != nil {
+		return errorOrIsNotFound(err, "failed to get the current RGW configuration period to see if it needs changed")
+	}
+
+	// this stages the current config changees and returns what the new period config will look like
+	// without committing the changes
+	stagedPeriod, err := runAdminCommand(c, true, "period", "update")
+	if err != nil {
+		return errorOrIsNotFound(err, "failed to stage the current RGW configuration period")
+	}
+
+	shouldCommit, err := periodWillChange(currentPeriod, stagedPeriod)
+	if err != nil {
+		return errors.Wrap(err, "failed to determine if the staged RGW configuration period is different from current")
+	}
+
+	if !shouldCommit {
+		nsName := fmt.Sprintf("%s/%s", c.clusterInfo.Namespace, c.Name)
+		logger.Debugf("not committing changes to RGW configuration period for CephObjectStore %q", nsName)
+		return nil
+	}
+	// don't expect json output since we don't intend to use the output from the command
+	_, err = runAdminCommand(c, false, "period", "update", "--commit")
+	if err != nil {
+		return errorOrIsNotFound(err, "failed to commit RGW configuration period changes")
+	}
+
+	return nil
+}
+
+// return true if the configuration period will change if the staged period is committed
+func periodWillChange(current, staged string) (bool, error) {
```

Comment:
I was imagining the period would be a simple id that could be compared. There is not a single field we can extract from the json to give a simpler comparison? Comparing the whole json and then ignoring certain fields seems like it could be prone to failure if in some configurations there are additional fields that weren't anticipated, or if things aren't sorted exactly the same, etc.

Reply:
Can we do something similar in createCephZoneGroup(), 
```
    // get period to see if master zone group exists yet
    output, err := object.RunAdminCommandNoMultisite(objContext, true, "period", "get", realmArg)
    if err != nil {
        if code, ok := exec.ExitStatus(err); ok && code == int(syscall.ENOENT) {
            return reconcile.Result{}, errors.Wrapf(err, "ceph period %q not found", zoneGroup.Spec.Realm)
        } else {
            return reconcile.Result{}, errors.Wrapf(err, "radosgw-admin period get failed with code %d", code)
        }
    }   

    // check if master zone group does not exist yet for period
    masterZoneGroup, err := decodeMasterZoneGroup(output)
    if err != nil {
        return reconcile.Result{}, errors.Wrapf(err, "failed to parse `radosgw-admin period get` output")
    }   
```
to resolve this issue, if the master zone group is updated with the value which we looking then the last period commit was successfully otherwise not.

Reply:
I don't know what I can say here that isn't commented in the code. 

There are things that report changes upcoming even when they won't change. And the one thing I actually expect to change -- `epoch` -- doesn't.

Perhaps @cbodley would offer whether we can request a deeper and more accurate feature from RGW itself for the future or if there is something I've missed that would be more useful. But I don't think we can do any better right now.

Reply:
#8879 has all of my pretty thorough notes about what the behavior of RGW is around the period.

Reply:
> Can we do something similar in createCephZoneGroup(),
> 
> ```
>     // get period to see if master zone group exists yet
>     output, err := object.RunAdminCommandNoMultisite(objContext, true, "period", "get", realmArg)
>     if err != nil {
>         if code, ok := exec.ExitStatus(err); ok && code == int(syscall.ENOENT) {
>             return reconcile.Result{}, errors.Wrapf(err, "ceph period %q not found", zoneGroup.Spec.Realm)
>         } else {
>             return reconcile.Result{}, errors.Wrapf(err, "radosgw-admin period get failed with code %d", code)
>         }
>     }   
> 
>     // check if master zone group does not exist yet for period
>     masterZoneGroup, err := decodeMasterZoneGroup(output)
>     if err != nil {
>         return reconcile.Result{}, errors.Wrapf(err, "failed to parse `radosgw-admin period get` output")
>     }   
> ```
> 
> to resolve this issue, if the master zone group is updated with the value which we looking then the last period commit was successfully otherwise not.

We possibly could look into whether there is a better implementation for `createCephZoneGroup()`, but I think this is a separate task/issue. That code doesn't seem to do anything idempotently that seems like an issue based on the problem statement from #8879

Reply:
As long as the sorting of the json doesn't change from one reconcile to the next and cause unnecessary period updates, it sounds fine to go with the current approach to ignore certain fields if rgw folks also agree. I'd rather err on the side of updating the period too often than not updating the period when it was needed. Object store reconciles really are  not very often after the initial object store creation.